ecorise_dashboard.py

```diff
@@ -60,37 +61,78 @@
+#     expanded = expanded.rename(lambda x:  x if x in theme_list else 'Other', axis=1)
+#     # Group 'Other' Columns together
+#     expanded = expanded.groupby(expanded.columns, axis=1).sum()
+#     # Merge the themes and expanded dataframes
+#     themes = pd.concat([themes, expanded], axis=1)
+#
+#     # Get Count of Programs per theme
+#     theme_count = pd.DataFrame(expanded.sum())
+#     theme_count.reset_index(inplace=True)
+#     theme_count.columns = ['Theme','Count']
+#     theme_count['Percent'] = round(100 * theme_count['Count'] / len(programs),0)
+#     theme_count = theme_count.astype({'Percent': int})
+#     theme_count = theme_count.sort_values(by=['Count'])
+#     return theme_count
+
+## Function to produce dataframe with: term, display_term
+def get_display_terms(filter_df, table, column):
+    filter_df = filter_df[filter_df['table_name'] == table]
+    filter_df = filter_df[filter_df['column_name'] == column]
+    filter_df = filter_df.loc[:, ['term','display_term']]
+    filter_df = filter_df.set_index('term')
+    return filter_df
+
+## Function to produce dataframe with: Environmental_Themes, Count, Percent
+def count_env_themes_for_programs(programs_df):
+    ## Get program names and environmental themes
+    program_themes_df = programs_df.loc[:, ['Program','Environmental_Themes']]
+    ## Split environmental themes from string into an array of strings on ", "
+    program_themes_df['Environmental_Themes'] = program_themes_df['Environmental_Themes'].str.split(', ')
+    ## Explode dataframe by environmental theme to get each theme from within an array into a row
+    program_themes_df = program_themes_df.explode('Environmental_Themes')
+#     theme_list = program_themes_df['Environmental_Themes'].unique()
+    ## Group by environmental theme counting the number of programs with each theme
+    theme_count_df = program_themes_df.groupby(['Environmental_Themes']).count()
+    ## Add column for percent of programs with the theme based on the number of program records
+    theme_count_df['Percent'] = round(100 * theme_count_df['Program'] / len(programs_df))
+    theme_count_df = theme_count_df.rename(columns={'Program': 'Count'})
+    theme_count_df = theme_count_df.astype({'Percent': int})
+#     print(theme_count_df)
+    return theme_count_df
+
+## Function to produce dataframe with: Environmental_Themes, Count, Percent, display_term, Label
+def prepare_env_themes_for_graph(programs_df, filter_df):
+    env_themes_df = count_env_themes_for_programs(programs_df)
+    env_theme_terms = get_display_terms(filter_df, 'Programs', 'Environmental_Themes')
+    ## Join environmental themes found in data with expected controlled terms and display terms
+    env_themes_df = env_themes_df.join(env_theme_terms)
+    ## Add column for label, using diplay term and percent value
+    env_themes_df['Label'] = env_themes_df['display_term'] + ' ' + env_themes_df['Percent'].astype(str) + '%'

```

Comment:
Concatenating strings requires this annoying type conversion with  `.astype(str)` or `str(value)`. [String formatting](https://pyformat.info) give you the ability to separate formatting from values. I'm going a little "overboard" with [named placeholders](https://pyformat.info/#named_placeholders), but it's clearer.
```suggestion
    env_themes_df['Label'] = "{display_term} {percentage}%".format(
        display_term=env_themes_df['display_term'],
        percentege=env_themes_df['Percent'],
    )

```

Reply:
I tried it that way in Lissa's Juptyer Notebook and recall it not working over each row in column, it did some weird combo of everything. So I gave up and went back to this format, but I'll double check what was happening.